sale_order_split_strategy/models/sale_order.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from collections import OrderedDict
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    split_strategy_id = fields.Many2one("sale.order.split.strategy")
+
+    def action_split(self):
+        orders_without_split = self.filtered(lambda o: not o.split_strategy_id)
+        if orders_without_split:
+            raise UserError(
+                _("Cannot split orders %s without any split strategy defined")
+                % ", ".join(orders_without_split.mapped("name"))
+            )
+        new_order_ids = []
+        for order in self:
+            copy_sections = order.split_strategy_id.copy_sections
+            copy_notes = order.split_strategy_id.copy_notes
+            lines_to_split = order.split_strategy_id._select_lines_to_split(order)
+            if not lines_to_split:
+                raise UserError(
+                    _(
+                        "Cannot split order %s according to its strategy because there are no matching lines"
+                    )
+                    % order.name
+                )
+            new_order = order.copy(self._prepare_order_copy_defaults())
+            # Prepare an OrderedDict from sale order lines and their sections
+            sections_dict = OrderedDict()
+            section_line_ids = None
+            for line in order.order_line.sorted():
```

Comment:
```suggestion
            for line in order.order_line:
```
As no key has been added, so default sort order

Reply:
We still need to make sure order lines are sorted using the default order, what is not necessarily the case with a recordset.

Reply:
What do you mean by default order ? A recordset is respecting _order property by default.

Reply:
Not necessarily :wink: 
```
>>> order = env["sale.order"].create({"partner_id": env["res.partner"].search([], limit=1).id})
>>> order
sale.order(27,)
>>> order.order_line
sale.order.line()
>>> product = env["product.product"].search([("sale_ok", "=", True)], limit=1)
>>> order.order_line.create({"order_id": order.id, "product_id": product.id, "product_uom_qty": 1})
sale.order.line(65,)
>>> order.order_line
sale.order.line(65,)
>>> order.order_line.sequence
10
>>> order.order_line.create({"order_id": order.id, "product_id": product.id, "product_uom_qty": 1, "sequence": 20})
sale.order.line(66,)
>>> order.order_line
sale.order.line(65, 66)
>>> order.order_line.create({"order_id": order.id, "product_id": product.id, "product_uom_qty": 1, "sequence": 5})
sale.order.line(67,)
>>> order.order_line
sale.order.line(65, 66, 67)
>>> order.order_line.sorted()
sale.order.line(67, 65, 66)
```

Reply:
As it seems linked to caching I'd rather invalidate the cache of `order_line` instead.